File: solutions/javascript/zebra-puzzle/2/zebra-puzzle.js
class House {
  colors        = ['red', 'ivory', 'green', 'yellow', 'blue'];
  nationalities = ['Englishman', 'Spaniard', 'Ukrainian', 'Japanese', 'Norwegian'];
  pets          = ['dog', 'snail', 'fox', 'horse', 'zebra'];
  beverages     = ['coffee', 'tea', 'milk', 'orange juice', 'water'];
  hobbies       = ['dancing', 'painting', 'reading', 'football', 'chess'];
}

class Solver {
  #houses;
  #updated;
  #validTrial;

  constructor() {
    this.#houses = [];

    for (let i = 0; i < 5; i++) {
      this.#houses.push(new House());
    }
  }

  solve() {
    // Known "knowns"
    this.#known(this.#houses, 2, 'beverages', 'milk');
    this.#known(this.#houses, 0, 'nationalities', 'Norwegian');
    this.#known(this.#houses, 1, 'colors', 'blue');

    // Eliminate as much as possible by applying each rule one at a time.
    this.#runRulesUntilNothingIsRemoved(this.#houses);

    // Brute force the rest
    this.#trialAndError(this.#houses);

    return this.#houses;
  }

  #runRulesUntilNothingIsRemoved(houses) {
    this.#updated = true;
    while (this.#updated) {
      this.#updated = false;
      this.#runRules(houses);
    }
  }

  #known(houses, index, key, value) {
    houses[index][key] = [value];
    for (let i = 0; i < 5; i++) {
      if (i !== index) {
        this.#remove(houses, i, key, value);
      }
    }
  }

  #remove(houses, index, key, value) {
    const position = houses[index][key].indexOf(value);

    if (position === -1) return;

    houses[index][key].splice(position, 1);
    this.#updated = true;

    if (houses[index][key].length === 0) {
      this.#validTrial = false;
    }
  }

  #runRules(houses) {
    for (let i = 0; i < 5; i++) {
      this.#same(houses, i, 'nationalities', 'Englishman', 'colors', 'red');
      this.#same(houses, i, 'nationalities', 'Spaniard', 'pets', 'dog');
      this.#same(houses, i, 'colors', 'green', 'beverages', 'coffee');
      this.#same(houses, i, 'nationalities', 'Ukrainian', 'beverages', 'tea');
      this.#leftRight(houses, i, 'colors', 'ivory', 'colors', 'green');
      this.#same(houses, i, 'pets', 'snail', 'hobbies', 'dancing');
      this.#same(houses, i, 'colors', 'yellow', 'hobbies', 'painting');
      this.#nextTo(houses, i, 'hobbies', 'reading', 'pets', 'fox');
      this.#nextTo(houses, i, 'hobbies', 'painting', 'pets', 'horse');
      this.#same(houses, i, 'hobbies', 'football', 'beverages', 'orange juice');
      this.#same(houses, i, 'nationalities', 'Japanese', 'hobbies', 'chess');
    }

    for (let i = 0; i < 5; i++) {
      for (let j = 0; j < 5; j++) {
        if (i === j) continue;
        this.#checkSolved(houses, j, 'colors', i);
        this.#checkSolved(houses, j, 'nationalities', i);
        this.#checkSolved(houses, j, 'pets', i);
        this.#checkSolved(houses, j, 'beverages', i);
        this.#checkSolved(houses, j, 'hobbies', i);
      }
    }
  }

  #checkSolved(houses, checkedIndex, key, indexToRemove) {
    if (!this.#isSolved(houses, checkedIndex, key)) return;

    this.#remove(
      houses,
      indexToRemove,
      key,
      houses[checkedIndex][key][0]
    );
  }

  #isSolved(houses, index, key) {
    return houses[index][key].length === 1
  }

  #isSolution(houses, index, key, value) {
    return (
         houses[index][key].length === 1
      && houses[index][key][0]     === value
    );
  }

  #isPossibility(houses, index, key, value) {
    return houses[index][key].includes(value);
  }

  #same(houses, index, key1, value1, key2, value2) {
    if (!this.#isPossibility(houses, index, key1, value1)) {
      this.#remove(houses, index, key2, value2);
    }

    if (!this.#isPossibility(houses, index, key2, value2)) {
      this.#remove(houses, index, key1, value1);
    }

    if (this.#isSolution(houses, index, key1, value1)) {
      this.#known(houses, index, key2, value2);
    }

    if (this.#isSolution(houses, index, key2, value2)) {
      this.#known(houses, index, key1, value1);
    }
  }

  #leftRight(houses, index, leftKey, leftValue, rightKey, rightValue) {
    this.#remove(houses, 0, rightKey, rightValue);
    this.#remove(houses, 4, leftKey, leftValue);

    if (index < 4) {
      if (!this.#isPossibility(houses, index, leftKey, leftValue)) {
        this.#remove(houses, index + 1, rightKey, rightValue);
      }

      if (this.#isSolution(houses, index, leftKey, leftValue)) {
        this.#known(houses, index + 1, rightKey, rightValue);
      }

      if (this.#isSolution(houses, index + 1, rightKey, rightValue)) {
        this.#known(houses, index, leftKey, leftValue);
      }
    }

    if (index > 0) {
      if (!this.#isPossibility(houses, index, rightKey, rightValue)) {
        this.#remove(houses, index - 1, leftKey, leftValue);
      }

      if (this.#isSolution(houses, index, rightKey, rightValue)) {
        this.#known(houses, index - 1, leftKey, leftValue);
      }

      if (this.#isSolution(houses, index - 1, leftKey, leftValue)) {
        this.#known(houses, index, rightKey, rightValue);
      }
    }
  }

  #nextTo(houses, index, key1, value1, key2, value2) {
    if (this.#isSolution(houses, index, key1, value1)) {
      if (index === 0) {
        this.#known(houses, index + 1, key2, value2);
      }

      if (index === 4) {
        this.#known(houses, index - 1, key2, value2);
      }

      this.#remove(houses, index, key2, value2);

      for (let i = index + 2; i < 5; i++) {
        this.#remove(houses, i, key2, value2);
      }

      for (let i = index - 2; i >= 0; i--) {
        this.#remove(houses, i, key2, value2);
      }
    }

    if (this.#isSolution(houses, index, key2, value2)) {
      if (index === 0) {
        this.#known(houses, index + 1, key1, value1);
      }

      if (index === 4) {
        this.#known(houses, index - 1, key1, value1);
      }

      this.#remove(houses, index, key1, value1);

      for (let i = index + 2; i < 5; i++) {
        this.#remove(houses, i, key1, value1);
      }

      for (let i = index - 2; i >= 0; i--) {
        this.#remove(houses, i, key1, value1);
      }
    }
  }

  #trialAndError(houses) {
    for (let i = 0; i < 5; i++) {
      this.#runTrial(houses, i, 'colors');
      this.#runTrial(houses, i, 'nationalities');
      this.#runTrial(houses, i, 'pets');
      this.#runTrial(houses, i, 'beverages');
      this.#runTrial(houses, i, 'hobbies');
    }
  }

  #runTrial(houses, index, key) {
    if (houses[index][key].length === 1) return;

    let trialHouses  = JSON.parse(JSON.stringify(houses));
    this.#validTrial = true;

    for (let trial = 0; trial < trialHouses[index][key].length; trial++) {
      this.#known(trialHouses, index, key, trialHouses[index][key][trial]);
      this.#runRulesUntilNothingIsRemoved(trialHouses);

      if (!this.#validTrial) {
        this.#remove(houses, index, key, trialHouses[index][key][trial]);
        this.#runRulesUntilNothingIsRemoved(houses);
      }
    }
  }
}

export class ZebraPuzzle {
  #houses;

  constructor(solver = new Solver()) {
    this.#houses = solver.solve();
  }

  waterDrinker() {
    return this.#houses.find(h => h.beverages[0] === 'water').nationalities[0];
  }

  zebraOwner() {
    return this.#houses.find(h => h.pets[0] === 'zebra').nationalities[0];
  }
}

File: solutions/ruby/zebra-puzzle/2/zebra_puzzle.rb
class ZebraPuzzle
  class << self
    def water_drinker
      houses.find { |h| h[:beverages].first == 'water' }[:nationalities].first
    end

    def zebra_owner
      houses.find { |h| h[:pets].first == 'zebra' }[:nationalities].first
    end

    private

    def houses
      ZebraPuzzleSolver.new.solve
    end
  end
end

class ZebraPuzzleSolver
  def initialize
    # Using the newer format with hobbies.
    # Ruby track is not yet using this at the time of writing this, but the
    # tests only care about nationalities, beverages, and pets.
    # Therefore, the other attributes can be arbitrarily substituted
    # so long as the rules stay consistent with the intended solution.
    house = {
      :colors        => ['red', 'ivory', 'green', 'yellow', 'blue'],
      :nationalities => ['Englishman', 'Spaniard', 'Ukrainian', 'Japanese', 'Norwegian'],
      :pets          => ['dog', 'snail', 'fox', 'horse', 'zebra'],
      :beverages     => ['coffee', 'tea', 'milk', 'orange juice', 'water'],
      :hobbies       => ['dancing', 'painting', 'reading', 'football', 'chess']
    }

    @houses      = []
    @updated     = true
    @valid_trial = true

    # Need a deep copy because we are using "process of elimination" to solve.
    (0..4).each { |i| @houses[i] = Marshal.load(Marshal.dump(house)) }
  end

  def solve
    # Known "knowns"
    known(@houses, 2, :beverages, 'milk')
    known(@houses, 0, :nationalities, 'Norwegian')
    known(@houses, 1, :colors, 'blue')

    # Eliminate as much as possible by applying each rule one at a time.
    run_rules_until_nothing_is_removed(@houses)

    # Brute force the rest by trying each possibility left.
    trial_and_error(@houses)

    @houses
  end

  private

  def known(houses, index, key, value)
    houses[index][key] = [value]

    houses.each_index do |i|
      remove!(houses, i, key, value) unless i == index
    end
  end

  def remove!(houses, index, key, value)
    return unless houses[index][key].include?(value)
    position = houses[index][key].index(value)
    houses[index][key].slice!(position)
    @updated = true
    @valid_trial = false if houses[index][key].empty?
  end

  def run_rules_until_nothing_is_removed(houses)
    @updated = true
    while @updated
      @updated = false
      run_rules(houses)
    end
  end

  def run_rules(houses)
    houses.each_index do |i|
      same(houses, i, :nationalities, 'Englishman', :colors, 'red')
      same(houses, i, :nationalities, 'Spaniard', :pets, 'dog')
      same(houses, i, :colors, 'green', :beverages, 'coffee')
      same(houses, i, :nationalities, 'Ukrainian', :beverages, 'tea')
      left_right(houses, i, :colors, 'ivory', :colors, 'green')
      same(houses, i, :pets, 'snail', :hobbies, 'dancing')
      same(houses, i, :colors, 'yellow', :hobbies, 'painting')
      next_to(houses, i, :hobbies, 'reading', :pets, 'fox')
      next_to(houses, i, :hobbies, 'painting', :pets, 'horse')
      same(houses, i, :hobbies, 'football', :beverages, 'orange juice')
      same(houses, i, :nationalities, 'Japanese', :hobbies, 'chess')
    end

    houses.each_index do |indexToRemove|
      houses.each_index do |indexToCheck|
        next if indexToCheck == indexToRemove
        check_solved(houses, indexToCheck, :colors,        indexToRemove)
        check_solved(houses, indexToCheck, :nationalities, indexToRemove)
        check_solved(houses, indexToCheck, :pets,          indexToRemove)
        check_solved(houses, indexToCheck, :beverages,     indexToRemove)
        check_solved(houses, indexToCheck, :hobbies,       indexToRemove)
      end
    end
  end

  def check_solved(houses, indexToCheck, key, indexToRemove)
    return unless solved?(houses, indexToCheck, key)
    remove!(houses, indexToRemove, key, houses[indexToCheck][key].first)
  end

  def solved?(houses, index, key)
    houses[index][key].length == 1
  end

  def same(houses, index, key_1, value_1, key_2, value_2)
    if !houses[index][key_1].include?(value_1)
      remove!(houses, index, key_2, value_2)
    end

    if !houses[index][key_2].include?(value_2)
      remove!(houses, index, key_1, value_1)
    end

    if solution?(houses, index, key_1, value_1)
      known(houses, index, key_2, value_2)
    end

    if solution?(houses, index, key_2, value_2)
      known(houses, index, key_1, value_1)
    end
  end

  def solution?(houses, index, key, value)
    houses[index][key].length == 1 && houses[index][key].first == value
  end

  def left_right(houses, index, left_key, left_value, right_key, right_value)
    remove!(houses, 0, right_key, right_value)
    remove!(houses, 4, left_key,  left_value )

    if index < 4
      if !houses[index][left_key].include?(left_value)
        remove!(houses, index + 1, right_key, right_value)
      end

      if solution?(houses, index, left_key, left_value)
        known(houses, index + 1, right_key, right_value)
      end

      if solution?(houses, index + 1, right_key, right_value)
        known(houses, index, left_key, left_value)
      end
    end

    if index > 0
      if !houses[index][right_key].include?(right_value)
        remove!(houses, index - 1, left_key, left_value)
      end

      if solution?(houses, index, right_key, right_value)
        known(houses, index - 1, left_key, left_value)
      end

      if solution?(houses, index - 1, left_key, left_value)
        known(houses, index, right_key, right_value)
      end
    end
  end

  def next_to(houses, index, key_1, value_1, key_2, value_2)
    if solution?(houses, index, key_1, value_1)
      known(houses, index + 1, key_2, value_2) if index.zero?
      known(houses, index - 1, key_2, value_2) if index == 4
      remove!(houses, index, key_2, value_2)
      (index + 2).upto(4)   { |i| remove!(houses, i, key_2, value_2) }
      (index - 2).downto(0) { |i| remove!(houses, i, key_2, value_2) }
    end

    if solution?(houses, index, key_2, value_2)
      known(houses, index + 1, key_1, value_1) if index.zero?
      known(houses, index - 1, key_1, value_1) if index == 4
      remove!(houses, index, key_1, value_1)
      (index + 2).upto(4)   { |i| remove!(houses, i, key_1, value_1) }
      (index - 2).downto(0) { |i| remove!(houses, i, key_1, value_1) }
    end
  end

  def trial_and_error(houses)
    houses.each_index do |i|
      run_trial(houses, i, :colors)
      run_trial(houses, i, :nationalities)
      run_trial(houses, i, :pets)
      run_trial(houses, i, :beverages)
      run_trial(houses, i, :hobbies)
    end
  end

  def run_trial(houses, index, key)
    return if solved?(houses, index, key)

    trial_houses = Marshal.load(Marshal.dump(houses))
    @valid_trial = true

    houses[index][key].each do |trial|
      known(trial_houses, index, key, trial)
      run_rules_until_nothing_is_removed(trial_houses)
      next if @valid_trial
      remove!(houses, index, key, trial)
      run_rules_until_nothing_is_removed(houses)
    end
  end
end
